internal/app/machined/pkg/controllers/runtime/kernel_param_spec.go

```diff
@@ -0,0 +1,146 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package runtime
+
+import (
+	"context"
+	"errors"
+	"os"
+	"strings"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"go.uber.org/zap"
+
+	"github.com/talos-systems/talos/pkg/kernel"
+	"github.com/talos-systems/talos/pkg/resources/runtime"
+)
+
+// KernelParamSpecController watches KernelParamSpecs, sets/resets kernel params.
+type KernelParamSpecController struct {
+	defaults map[string]string
+	state    map[string]string
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *KernelParamSpecController) Name() string {
+	return "runtime.KernelParamSpecController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *KernelParamSpecController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: runtime.NamespaceName,
+			Type:      runtime.KernelParamSpecType,
+			Kind:      controller.InputStrong,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *KernelParamSpecController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: runtime.KernelParamStatusType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//nolint:gocyclo
+func (ctrl *KernelParamSpecController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	if ctrl.state == nil {
+		ctrl.state = map[string]string{}
+	}
+
+	if ctrl.defaults == nil {
+		ctrl.defaults = map[string]string{}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+			list, err := r.List(ctx, resource.NewMetadata(runtime.NamespaceName, runtime.KernelParamSpecType, "", resource.VersionUndefined))
+			if err != nil {
+				return err
+			}
+
+			touchedIDs := map[string]struct{}{}
+
+			for _, item := range list.Items {
+				spec := item.(*runtime.KernelParamSpec).TypedSpec()
+
+				if err = ctrl.updateKernelParam(ctx, r, item.Metadata().ID(), spec.Value); err != nil {
```

Comment:
one thing we _might_ want to do here is to make this loop accumulate errors in `multierror` and return error only when done with all the sysctls

the reason is that some sysctls might have dependency on each other, as we don't enforce any order here explicitly (except for alphabetical), it might be that one sysctl can be written only after other sysctl is written or reset back to its default value, so I would suggest to run both loops and only report error afterwards